Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp

```diff
@@ -0,0 +1,50 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Core/HW/HSP/HSP_DeviceARAMExpansion.h"
+
+#include "Common/ChunkFile.h"
+#include "Common/Logging/Log.h"
+#include "Common/MathUtil.h"
+#include "Common/MemoryUtil.h"
+#include "Common/Swap.h"
+
+#include "Core/ConfigManager.h"
+#include "Core/HW/DSP.h"
+#include "Core/HW/HSP/HSP.h"
+#include "Core/HW/ProcessorInterface.h"
+
+#include <cstring>
+
+namespace HSP
+{
+CHSPDevice_ARAMExpansion::CHSPDevice_ARAMExpansion(HSPDevices device) : IHSPDevice(device)
+{
+  m_size = MathUtil::NextPowerOf2(SConfig::GetInstance().m_ARAMExpansion);
+  m_mask = m_size - 1;
+  m_ptr = static_cast<u8*>(Common::AllocateMemoryPages(m_size));
+}
+
+CHSPDevice_ARAMExpansion::~CHSPDevice_ARAMExpansion()
+{
+  Common::FreeMemoryPages(m_ptr, m_size);
+  m_ptr = nullptr;
+}
+
+u64 CHSPDevice_ARAMExpansion::Read(u32 address)
+{
+  return Common::swap64(*(u64*)&m_ptr[address & m_mask]);
```

Comment:
This is undefined behavior.

```cpp
u64 value;
std::memcpy(&value, &m_ptr[address & mask], sizeof(value));
return Common::swap64(value);
```

Reply:
Or use the handy [`Common::BitCast`](https://github.com/dolphin-emu/dolphin/blob/3b4486bad3f0e44c3a572bba411879a634ece325/Source/Core/Common/BitUtils.h#L189)